File: java/PleskApiClient.java
// Copyright 1999-2016. Parallels IP Holdings GmbH. All Rights Reserved.

import java.net.*;
import java.io.*;
import javax.net.ssl.*;
import java.security.cert.X509Certificate;

class PleskApiClient {

    private String host;
    private String login;
    private String password;
    private String secretKey;

    public PleskApiClient(String host) {
        this.host = host;
    }

    public void setCredentials(String login, String password) {
        this.login = login;
        this.password = password;
    }

    public void setSecretKey(String secretKey) {
        this.secretKey = secretKey;
    }

    public String request(String request) throws Exception {
        this.setTrustAllSslCertificates();

        URL url = new URL("https://" + this.host + ":8443/enterprise/control/agent.php");
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();

        connection.setRequestMethod("POST");
        connection.setRequestProperty("Content-Type", "text/xml");
        connection.setRequestProperty("HTTP_PRETTY_PRINT", "TRUE");

        if (null != this.secretKey) {
            connection.setRequestProperty("KEY", this.secretKey);
        } else {
            connection.setRequestProperty("HTTP_AUTH_LOGIN", this.login);
            connection.setRequestProperty("HTTP_AUTH_PASSWD", this.password);
        }

        connection.setDoOutput(true);

        DataOutputStream writer = new DataOutputStream(connection.getOutputStream());
        writer.writeBytes(request);
        writer.flush();
        writer.close();

        BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));

        String response = "";
        String inputLine;

        while ((inputLine = reader.readLine()) != null) {
            response += inputLine + "\n";
        }

        reader.close();

        return response;
    }

    private void setTrustAllSslCertificates() throws Exception {
        TrustManager[] trustAllCerts = new TrustManager[] {new X509TrustManager() {
                public java.security.cert.X509Certificate[] getAcceptedIssuers() {
                    return null;
                }
                public void checkClientTrusted(X509Certificate[] certs, String authType) {
                }
                public void checkServerTrusted(X509Certificate[] certs, String authType) {
                }
            }
        };

        SSLContext sc = SSLContext.getInstance("SSL");
        sc.init(null, trustAllCerts, new java.security.SecureRandom());
        HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());

        HostnameVerifier allHostsValid = new HostnameVerifier() {
            public boolean verify(String hostname, SSLSession session) {
                return true;
            }
        };

        HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
    }

}

File: cpp/plesk_api_client.cpp
// Copyright 1999-2016. Parallels IP Holdings GmbH. All Rights Reserved.

#include <iostream>
#include <curl/curl.h>
#include "plesk_api_client.h"

PleskApiClient::PleskApiClient(const string& host)
{
    m_host = host;
}

void PleskApiClient::setCredentials(const string& login, const string& password)
{
    m_login = login;
    m_password = password;
}

size_t writeToString(void *ptr, size_t size, size_t count, void *stream)
{
    ((string*)stream)->append((char*)ptr, 0, size*count);
    return size*count;
}

void PleskApiClient::request(const string& request, string& response)
{
    CURL *curl;
    CURLcode res;
    struct curl_slist *headers = NULL;

    curl = curl_easy_init();

    if (curl) {
        headers = curl_slist_append(headers, "Content-type: text/xml");
        headers = curl_slist_append(headers, "HTTP_PRETTY_PRINT: TRUE");
        headers = curl_slist_append(headers, ("HTTP_AUTH_LOGIN: " + m_login).c_str());
        headers = curl_slist_append(headers, ("HTTP_AUTH_PASSWD: " + m_password).c_str());

        curl_easy_setopt(curl, CURLOPT_URL, string(string("https://") + m_host + string(":8443/enterprise/control/agent.php")).c_str());
        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, request.c_str());
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writeToString);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);

        res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
        }
        curl_easy_cleanup(curl);
        curl_slist_free_all(headers);
    }
}
